### PR TITLE
Fixes DB DDL related race condition

### PR DESCRIPTION
**Pull request by nyarly**
* Wraps DDL modifications in a Once
* Reports on errors from the DDL change
* "scoops" out NameCaches instead of reusing the constructor

### CODE REVIEW

**Comment by sous-machine-account** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please include a CHANGELOG entry.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: ext/docker/image_mapping.go [hunk 1]
```diff
@@ -458,37 +459,41 @@ func semverEqual(a, b string) (bool, error) {
...
-		err = nil
-		repos := nc.captureRepos(db)
-
-		nc.clobber(db)
-
-		for _, cmd := range schema {
-			if err := sqlExec(db, cmd); err != nil {
-				return errors.Wrapf(err, "groom DB/create: %v", db)
+	nc.groomOnce.Do(func() {
+		queryErr := db.QueryRow("select value from _database_metadata_ where name = 'fingerprint';").Scan(&tgp)
```
Comment (richardhopton) [inline ext/docker/image_mapping.go]: I've found this check rarely fixes race conditions.  A better option is to try the insert and if it fails then look to see if the document is there and respond success if so.

**Comment by richardhopton** [comment]
Did you mean to squash your commit?

File: ext/docker/image_mapping.go (hunk 1 above)
Comment (nyarly) [reply ext/docker/image_mapping.go]: That's not what this check is for. It confirms that the database-as-is has the right schema; if it does we don't have to fix it. The fingerprint is recorded as the last step, and if it's generated as a digest of the DDL code.

Also: insert-then-select narrows the window but still doesn't close the race condition problem. What you want is upsert, which is a fairly late addition in SQL (e.g. Postgres 9.4, I think, allows for it?) We fake it here with some unique indexes and some insert...or update lines.

**Comment by nyarly** [comment]
I'm okay without squashing.

**Pull request merged**